libs/header/sync/sync.go

```diff
@@ -240,44 +235,103 @@ func (s *Syncer[H]) doSync(ctx context.Context, fromHead, toHead H) (err error)
 
 // processHeaders gets and stores headers starting at the given 'from' height up to 'to' height -
 // [from:to]
-func (s *Syncer[H]) processHeaders(ctx context.Context, from, to uint64) (int, error) {
-	headers, err := s.findHeaders(ctx, from, to)
+func (s *Syncer[H]) processHeaders(ctx context.Context, fromHeader H, to uint64) error {
+	headers, err := s.requestHeaders(ctx, fromHeader, to)
 	if err != nil {
-		return 0, err
+		return err
+	}
+	_, err = s.store.Append(ctx, headers...)
+	return err
+}
+
+// requestHeaders checks headers in pending cache that apply to the requested range.
+// If some headers are missing, it starts requesting them from the network.
+// All headers that will be received are verified to be contiguous.
+func (s *Syncer[H]) requestHeaders(
+	ctx context.Context,
+	fromHeader H,
+	to uint64,
+) ([]H, error) {
+	amount := to - uint64(fromHeader.Height())
+	cached, ok := s.checkCache(fromHeader, to)
+	if !ok {
+		// request full range if cache is empty
+		return s.findHeaders(ctx, fromHeader, to)
 	}
 
-	return s.store.Append(ctx, headers...)
+	out := make([]H, 0, amount)
+	for _, headers := range cached {
+		if fromHeader.Height()+1 == headers[0].Height() {
+			// apply cache
+			out = append(out, headers...)
+			// set new header to count from
+			fromHeader = out[len(out)-1]
+			continue
+		}
+		// make an external request
+		h, err := s.findHeaders(ctx, fromHeader, uint64(headers[0].Height())-1)
+		if err != nil {
+			return nil, err
+		}
+		// apply received headers + headers from the range
+		out = append(out, append(h, headers...)...)
+		fromHeader = out[len(out)-1]
+	}
+
+	// ensure that we have all requested headers
+	if uint64(len(out)) == amount {
+		return out, nil
+	}
+	// make one more external request in case if `to` is bigger than the
+	// last cached header
+	h, err := s.findHeaders(ctx, fromHeader, to)
+	if err != nil {
+		return nil, err
+	}
+	return append(out, h...), nil
 }
 
-// findHeaders gets headers from either remote peers or from local cache of headers received by
-// PubSub - [from:to]
-func (s *Syncer[H]) findHeaders(ctx context.Context, from, to uint64) ([]H, error) {
-	amount := to - from + 1 // + 1 to include 'to' height as well
-	if amount > s.Params.MaxRequestSize {
-		to, amount = from+s.Params.MaxRequestSize, s.Params.MaxRequestSize
+// checkCache returns all headers sub-ranges of headers that could be found in between the requested range.
+func (s *Syncer[H]) checkCache(fromHeader H, to uint64) ([][]H, bool) {
+	r := s.pending.FindAllRangesWithin(uint64(fromHeader.Height()+1), to) // start looking for headers for the next height
+	if len(r) == 0 {
+		return nil, false
 	}
 
+	out := make([][]H, len(r))
+	for i, headersRange := range r {
+		cached, _ := headersRange.Before(to)
+		out[i] = append(out[i], cached...)
+	}
+	return out, len(out) > 0
+}
+
+// findHeaders requests headers from the network -> (fromHeader.Height;to].
+func (s *Syncer[H]) findHeaders(
```

Comment:
I feel like maybe this method should be the one called `requestHeaders` and the current `requestHeaders` could be called the more generic `findHeaders` or `getHeaders`, since it uses both a cache get and network get

Reply:
`processHeaders `should be fine I think.